pinot-controller/src/main/java/org/apache/pinot/controller/util/TableMetadataReader.java

```diff
@@ -127,40 +128,108 @@ private TableReloadJsonResponse processSegmentMetadataReloadResponse(
 
   /**
    * This api takes in list of segments for which we need the metadata.
+   * This calls the server to get the metadata for all segments instead of making a call per segment.
    */
   public JsonNode getSegmentsMetadata(String tableNameWithType, List<String> columns, Set<String> segmentsToInclude,
       int timeoutMs)
       throws InvalidConfigException, IOException {
     return getSegmentsMetadataInternal(tableNameWithType, columns, segmentsToInclude, timeoutMs);
   }
 
+  /**
+   *   Common helper used by both the new (table-level) and legacy
+   *   (segment-level) endpoints.
+   */
+  private JsonNode fetchAndAggregateMetadata(List<String> urls,
+      BiMap<String, String> endpoints,
+      boolean perSegmentJson,
+      String tableNameWithType,
+      int timeoutMs)
+      throws InvalidConfigException, IOException {
+
+    CompletionServiceHelper cs =
+        new CompletionServiceHelper(_executor, _connectionManager, endpoints);
+    CompletionServiceHelper.CompletionServiceResponse resp =
+        cs.doMultiGetRequest(urls, tableNameWithType, perSegmentJson, timeoutMs);
+
+    if (resp._failedResponseCount == urls.size()) {
+      throw new InvalidConfigException("All requests to server instances failed.");
+    }
+
+    Map<String, JsonNode> aggregated = new HashMap<>();
```

Comment:
If we are going to return a json object in the end, we can probably directly create an `ObjectNode` here